docs/proposals/20220717-resourcedistribution-generator.md

```diff
@@ -0,0 +1,293 @@
+      - [Target Namespaces Rules](target-namespaces-rules)
+      - [Behavior](#behavior)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+`ResourceDistribution` is a resource provided by OpenKruise that is used to distribute and synchronize `ConfigMap`, `Secret`, and other resources across multiple namespaces. Using `ResourceDistribution` requires a full description of the resource being distributed, and manually modifying and maintaining the YAML file format is very difficult. `ConfigMapGenerator` avoids this problem by importing data directly from a file to create a confimap. We need a similar tool to create a `ResourceDistribution` by importing data directly from a file. Kustomize supports custom plugin to help us do this.
+
+## Proposal
+
+### API Definition
+
+```go
+package main
+
+type plugin struct{
+	rf        *resmap.Factory
+	ldr       ifc.Loader
+	Name      string `json:"name,omitempty" yaml:"name,omitempty"`
+  ResourceArgs `json:",inline,omitempty" yaml:",inline,omitempty"`
+  TargetsArgs `json:",inline,omitempty" yaml:",inline,omitempty"`	
+  
+  // Options for the resourcedistribution.
+  // GeneratorOptions same as confgimap and secret generator Options
+  Options *GeneratorOptions `json:"options,omitempty" yaml:"options,omitempty"`
+
+  // Behavior of generated resource, must be one of:
+	//   'create': create a new one
+	//   'replace': replace the existing one
+	//   'merge': merge with the existing one
+	Behavior string `json:"behavior,omitempty" yaml:"behavior,omitempty"`
+}
+
+// ResourceArgs contain arguments for the resource to be distributed.
+type ResourceArgs struct {
+  // Name of the resource to be distributed.
+	DistributionName string `json:"distributionName,omitempty" yaml:"distributionName,omitempty"`
+  
+  // KvPairSources defines places to obtain key value pairs.
+  // same as confgimap and secret generator KvPairSources
+	KvPairSources `json:",inline,omitempty" yaml:",inline,omitempty"`
+  
+  // Optinos for the resource to be distributed.
+  // GeneratorOptions same as confgimap and secret generator Options
+	DistributionOptions *GeneratorOptions `json:"distributionOptions,omitempty" yaml:"distributionOptions,omitempty"`
+}
+
+// TargetsArgs defines places to obtain target namespace args.
+type TargetsArgs struct {
+  // AllNamespaces if true distribute all namespaces
+  AllNamespaces bool `json:"allNamespaces,omitempty" yaml:"allNamespaces,omitempty"`
+  
+  // ExcludedNamespaces is a list of excluded namespaces name.
+  ExcludedNamespaces []string `json:"excludedNamespaces,omitempty" yaml:"excludedNamespaces,omitempty"`
+  
+  // IncludedNamespaces is a list of included namespaces name.
+  IncludedNamespaces []string `json:"includedNamespaces,omitempty" yaml:"includedNamespaces,omitempty"`
+ 
+  // NamespaceLabelSelector for the generator.
+  NamespaceLabelSelector *namespaceLabelSelector `json:"namespaceLabelSelector,omitempty" yaml:"namespaceLabelSelector,omitempty"`
```

Comment:
How about using `NamespaceLabelSelector metav1.LabelSelector `json:"namespaceLabelSelector,omitempty"`` directly?

Reply:
Good idea. I will modify it.